src/main/java/uk/gov/ch/developer/docs/controller/developer/SignInController.java

```diff
@@ -0,0 +1,231 @@
+     * @param request
+     * @param response
+     * @param companyNumber The company number a user is viewing
+     * @param force Boolean determining whether to force re-authentication
+     * @throws IOException
+     */
+    protected void redirectForAuth(Session session, HttpServletRequest request,
+            HttpServletResponse response, String companyNumber, boolean force) throws IOException {
+
+        // Find the original requested url
+        StringBuilder originalRequestUrl = new StringBuilder(request.getRequestURL().toString());
+        String queryString = request.getQueryString();
+        if (queryString != null)
+            originalRequestUrl.append("?").append(queryString);
+
+        // Set the scope
+        String scope = null;
+        if (companyNumber != null)
+            scope = OAUTH_COMPANY_SCOPE_PREFIX + companyNumber;
+
+        // Generate and store a nonce in the session
+        Session sessionToUpdate = session;
+
+        if (sessionToUpdate == null) {
+            sessionToUpdate = new SessionImpl();
+        }
+
+        String nonce = generateNonce();
+        sessionToUpdate.getData().put(SessionKeys.NONCE.getKey(), nonce);
+        
+//        //Store the CHS session
+//        chSession.store();
+
+        // Build oauth uri and redirect
+        String authoriseUri;
+        if (force) {
+            authoriseUri = createAuthoriseURIWithForceAndHint(originalRequestUrl.toString(), scope,
+                    nonce, getEmailFromSession(session));
+        } else {
+            authoriseUri = createAuthoriseURI(originalRequestUrl.toString(), scope, nonce);
+        }
+
+        response.sendRedirect(authoriseUri);
+    }
+
+    /**
+     * Retrieves email from a session if present, else returns an empty string
+     * 
+     * @param session User session from which to retrieve the email address
+     * @return email
+     */
+    private String getEmailFromSession(Session session) {
+
+        String email = "";
+
+        SignInInfo signInInfo = new SignInInfo();
+        if (session != null) {
+            signInInfo = session.getSignInInfo();
+        }
+
+        UserProfile userProfile = signInInfo.getUserProfile();
+        if (userProfile != null)
+            email = userProfile.getEmail();
+
+        return email;
+    }
+
+    /**
+     * Constructs the authorisation URI with additional force and hint
+     * 
+     * @param originalRequestUri Original URI from which to redirect
+     * @param scope Scope of the request
+     * @param nonce
+     * @param email Email address from the session, empty if not present
+     * @return Authorisation URI
+     */
+    protected String createAuthoriseURIWithForceAndHint(String originalRequestUri, String scope,
+            String nonce, String email) {
+
+        StringBuilder sb = new StringBuilder(createAuthoriseURI(originalRequestUri, scope, nonce));
+        sb.append("&reauthenticate=force");
```

Comment:
Stringbuilder append returns itself so we could chain these methods.

Reply:
We could, I see no benefit other than syntactical?

Reply:
Isn't that a worthwhile benefit if changes need to happen anyway?